In [85]:
import requests
import json
import re
from json import loads
import datetime
import time

url = "https://apis.openapi.sk.com/tmap/routes"

# 출발
start_name = "휘경2동 주민센터"
start_X = 127.06868
start_Y = 37.589912

# 도착
end_name = "용두동 사거리"
end_X = 127.031013
end_Y = 37.577686

headers = {
    "appkey": "l7xx66f8e738470b4bf79d908f1dca241eff",
    "version": "1",
    "callback": "",
    "Content-Type": 'totalValue="2"'
}

In [6]:
payload = {
    "startName" : "출발 : " + str(start_name),
    "startX" : str(start_X),
    "startY" : str(start_Y),
    "endName" : "도착 : " + str(end_name),
    "endX" : str(end_X),
    "endY" : str(end_Y)
}

In [7]:
r = requests.post(url, json=payload, headers=headers)
 
jsonObj = json.loads(r.text)

jsonObj

{'type': 'FeatureCollection',
 'features': [{'type': 'Feature',
   'geometry': {'type': 'Point',
    'coordinates': [127.06869825291737, 37.58980637600597]},
   'properties': {'totalDistance': 4485,
    'totalTime': 1151,
    'totalFare': 0,
    'taxiFare': 7600,
    'index': 0,
    'pointIndex': 0,
    'name': '',
    'description': '외대역동로 을 따라  방면으로 251m 이동',
    'nextRoadName': '외대역동로',
    'turnType': 200,
    'pointType': 'S'}},
  {'type': 'Feature',
   'geometry': {'type': 'LineString',
    'coordinates': [[127.06869825291737, 37.58980637600597],
     [127.06865936780824, 37.58979804294082],
     [127.06859826226419, 37.58979804183977],
     [127.06853437824702, 37.589831370146584],
     [127.06840938639388, 37.58994524354228],
     [127.0683316130934, 37.59003689815016],
     [127.06829550377032, 37.59008966914121],
     [127.06823439506499, 37.59020076623312],
     [127.06806495515006, 37.590584051945754],
     [127.06798718129637, 37.5906951487373],
     [127.06774552992741, 3

In [19]:
print(datetime.timedelta(0, int(jsonObj['features'][0]['properties']['totalTime'])))
print(jsonObj['features'][0]['properties']['totalTime'])

0:19:11
1151


## 따릉이 대여소

In [2]:
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import warnings

warnings.filterwarnings('ignore')

mpl.rcParams['axes.unicode_minus'] = False
plt.rcParams["font.family"] = 'NanumGothic'


In [3]:
Bike_Stop = pd.read_excel('./raw data/대여소/서울특별시 공공자전거 대여소 정보(19.12.9).xlsx')
Bike_Stop_ID = Bike_Stop.groupby('대여소_구')['대여소ID'].unique().loc['동대문구']
#Bike_Stop_ID = np.append(Bike_Stop_ID, 664)
Bike_Stop_ID = Bike_Stop_ID.astype(np.int64)
Bike_Stop_ID

array([600, 601, 602, 604, 605, 606, 607, 608, 609, 610, 612, 613, 614,
       615, 616, 617, 621, 622, 623, 624, 625, 626, 627, 628, 630, 631,
       633, 634, 635, 636, 637, 638, 639, 640, 641, 642, 643, 646, 647,
       648, 650, 651, 652, 654, 656, 657, 658, 659, 660, 661, 663],
      dtype=int64)

In [3]:
idx_id = Bike_Stop[Bike_Stop['대여소ID'].isin(Bike_Stop_ID)].index
idx_id

Int64Index([399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411,
            412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422, 423, 424,
            425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436, 437,
            438, 439, 440, 441, 442, 443, 444, 445, 446, 447, 448, 449],
           dtype='int64')

In [4]:
Bike_Stop_POS = Bike_Stop.loc[idx_id, ['대여소ID', '경도', '위도']].reset_index(drop = True)
Bike_Stop_POS = Bike_Stop_POS.astype({'대여소ID': np.int64})

In [5]:
Bike_Stop_POS

대여소ID          경도         위도
0     600  127.068680  37.589912
1     601  127.037361  37.575947
2     602  127.071388  37.572174
3     604  127.055962  37.569656
4     605  127.026497  37.574200
5     606  127.070290  37.584625
6     607  127.067863  37.602711
7     608  127.055344  37.581310
8     609  127.037361  37.587791
9     610  127.052162  37.573086
10    612  127.030815  37.573666
11    613  127.023468  37.575272
12    614  127.031013  37.577686
13    615  127.035110  37.576382
14    616  127.054367  37.582561
15    617  127.057693  37.574203
16    621  127.053391  37.574718
17    622  127.057831  37.577793
18    623  127.053856  37.583698
19    624  127.045891  37.574188
20    625  127.057175  37.568192
21    626  127.070969  37.561153
22    627  127.071907  37.578632
23    628  127.067543  37.586815
24    630  127.040306  37.574852
25    631  127.051277  37.568184
26    633  127.044823  37.580406
27    634  127.059830  37.596020
28    635  127.052773  37.587517
29    636  127.042587  37.590900
30    637  127.045792  37.591614
31    638  127.060974  37.583008
32    639  127.060951  37.585197
33    640  127.044495  37.582455
34    641  127.038536  37.573753
35    642  127.048584  37.570156
36    643  127.074417  37.566971
37    646  127.064377  37.561867
38    647  127.067268  37.602798
39    648  127.062424  37.566700
40    650  127.067741  37.591534
41    651  127.048279  37.582748
42    652  127.065941  37.576530
43    654  127.052109  37.577621
44    656  127.043137  37.587238
45    657  127.048370  37.575760
46    658  127.066147  37.572948
47    659  127.033386  37.578350
48    660  127.037628  37.577999
49    661  127.051964  37.594601
50    663  127.063026  37.594891

In [6]:
Bike_Stop_Dic = {}
for idx, i in zip(Bike_Stop_ID, range(51)):
    Bike_Stop_Dic[str(idx)] = {}
    Bike_Stop_Dic[str(idx)]['x'] = str(Bike_Stop_POS.iloc[i,1])
    Bike_Stop_Dic[str(idx)]['y'] = str(Bike_Stop_POS.iloc[i,2])
    
Bike_Stop_Dic

{'600': {'x': '127.06868', 'y': '37.589912'},
 '601': {'x': '127.037361', 'y': '37.575947'},
 '602': {'x': '127.071388', 'y': '37.572174'},
 '604': {'x': '127.055962', 'y': '37.569656'},
 '605': {'x': '127.026497', 'y': '37.5742'},
 '606': {'x': '127.07029', 'y': '37.584625'},
 '607': {'x': '127.067863', 'y': '37.602711'},
 '608': {'x': '127.055344', 'y': '37.58131'},
 '609': {'x': '127.037361', 'y': '37.587791'},
 '610': {'x': '127.052162', 'y': '37.573086'},
 '612': {'x': '127.030815', 'y': '37.573666'},
 '613': {'x': '127.023468', 'y': '37.575272'},
 '614': {'x': '127.031013', 'y': '37.577686'},
 '615': {'x': '127.03511', 'y': '37.576382'},
 '616': {'x': '127.054367', 'y': '37.582561'},
 '617': {'x': '127.057693', 'y': '37.574203'},
 '621': {'x': '127.053391', 'y': '37.574718'},
 '622': {'x': '127.057831', 'y': '37.577793'},
 '623': {'x': '127.053856', 'y': '37.583698'},
 '624': {'x': '127.045891', 'y': '37.574188'},
 '625': {'x': '127.057175', 'y': '37.568192'},
 '626': {'x': '127.

In [7]:
import pickle
with open('./time/Bike_Stop_POS.pickle','wb') as fw:
    pickle.dump(Bike_Stop_Dic, fw)

### 대여소간 소요 시간
- 무료 key
    - l7xx66f8e738470b4bf79d908f1dca241eff
- 유료 key
    - l7xx9ed2ada547c14e6088e9e681fce631f7

In [103]:
def getTotalTime(start_id, end_id):
    url = "https://apis.openapi.sk.com/tmap/routes"

    headers = {
        "appkey": "l7xx9ed2ada547c14e6088e9e681fce631f7",
        "version": "1",
        "callback": ""
    }
    
    payload = {
        "startX" : start_id['x'],
        "startY" : start_id['y'],
        "endX" : end_id['x'],
        "endY" : end_id['y']
    }
    
    r = requests.post(url, json=payload, headers=headers)

    jsonObj = json.loads(r.text)

    time.sleep(0.3) # {'error': {'id': '403', 'category': 'gw', 'code': '8101', 'message': '초당 처리 건수를 초과했습니다.'}} 오류 해결 위해서 delay 부여
    
    #return int(jsonObj['features'][0]['properties']['totalTime'])
    try:
        return int(jsonObj['features'][0]['properties']['totalTime'])
    except:
        print(jsonObj)

In [104]:
Bike_Stop_Time = {}
for ind_s in Bike_Stop_ID:
    Bike_Stop_Time[str(ind_s)] = {}
    for ind_e in Bike_Stop_ID:
        Bike_Stop_Time[str(ind_s)][str(ind_e)] = getTotalTime(Bike_Stop_Dic[str(ind_s)], Bike_Stop_Dic[str(ind_e)])
    print(str(ind_s) + "번 대여소 완료")


600번 대여소 완료
601번 대여소 완료
602번 대여소 완료
604번 대여소 완료
605번 대여소 완료
606번 대여소 완료
607번 대여소 완료
608번 대여소 완료
609번 대여소 완료
610번 대여소 완료
612번 대여소 완료
613번 대여소 완료
614번 대여소 완료
615번 대여소 완료
616번 대여소 완료
617번 대여소 완료
621번 대여소 완료
622번 대여소 완료
623번 대여소 완료
624번 대여소 완료
625번 대여소 완료
626번 대여소 완료
627번 대여소 완료
628번 대여소 완료
630번 대여소 완료
631번 대여소 완료
633번 대여소 완료
634번 대여소 완료
635번 대여소 완료
636번 대여소 완료
637번 대여소 완료
638번 대여소 완료
639번 대여소 완료
640번 대여소 완료
641번 대여소 완료
642번 대여소 완료
643번 대여소 완료
646번 대여소 완료
647번 대여소 완료
648번 대여소 완료
650번 대여소 완료
651번 대여소 완료
652번 대여소 완료
654번 대여소 완료
656번 대여소 완료
657번 대여소 완료
658번 대여소 완료
659번 대여소 완료
660번 대여소 완료
661번 대여소 완료
663번 대여소 완료


In [111]:
Bike_Stop_Time

{'600': {'600': 0,
  '601': 925,
  '602': 757,
  '604': 1086,
  '605': 1053,
  '606': 485,
  '607': 702,
  '608': 781,
  '609': 886,
  '610': 1168,
  '612': 1178,
  '613': 1462,
  '614': 1013,
  '615': 981,
  '616': 730,
  '617': 889,
  '621': 856,
  '622': 731,
  '623': 766,
  '624': 903,
  '625': 1020,
  '626': 830,
  '627': 668,
  '628': 542,
  '630': 1058,
  '631': 1118,
  '633': 592,
  '634': 414,
  '635': 270,
  '636': 551,
  '637': 533,
  '638': 908,
  '639': 267,
  '640': 453,
  '641': 880,
  '642': 994,
  '643': 701,
  '646': 965,
  '647': 799,
  '648': 1298,
  '650': 114,
  '651': 347,
  '652': 708,
  '654': 917,
  '656': 467,
  '657': 878,
  '658': 728,
  '659': 765,
  '660': 1119,
  '661': 448,
  '663': 253},
 '601': {'600': 1690,
  '601': 0,
  '602': 1401,
  '604': 852,
  '605': 243,
  '606': 1932,
  '607': 1940,
  '608': 721,
  '609': 755,
  '610': 808,
  '612': 413,
  '613': 680,
  '614': 187,
  '615': 99,
  '616': 670,
  '617': 856,
  '621': 765,
  '622': 1359,
  '623':

In [106]:
import pickle
with open('./time/Bike_Stop_Time.pickle','wb') as fw:
    pickle.dump(Bike_Stop_Time, fw)

In [110]:
with open('./time/Bike_Stop_Time.pickle', 'rb') as fr:
    user_loaded = pickle.load(fr)
    
print(user_loaded['663']['661'])

660


### ./time/Bike_Stop_Time.pickle 로 저장함

In [113]:
print(max(user_loaded.values().values()))

AttributeError: 'dict_values' object has no attribute 'values'

In [118]:
max(user_loaded['600'].values())

1462

In [121]:
temp = {}
for k in Bike_Stop_ID:
    temp[str(k)] = max(user_loaded[str(k)].values())

In [131]:
max(temp.values())

2814

In [126]:
print(datetime.timedelta(0, max(temp.values())))

0:46:54


In [171]:
temp1 = {}
for k, l in zip(Bike_Stop_ID, range(51)):
    #temp1[str(k)] = min(list(user_loaded[str(k)].values()).remove(0))
    #print(del list(user_loaded[str(k)].values())[l])
    t = list(user_loaded[str(k)].values())
    del t[l]
    #print(t)
    #print(min(t))
    temp1[str(k)] = min(t)

In [173]:
min(temp1.values())

35

In [174]:
print(datetime.timedelta(0, min(temp1.values())))

0:00:35


In [172]:
temp1

{'600': 114,
 '601': 99,
 '602': 195,
 '604': 73,
 '605': 469,
 '606': 190,
 '607': 234,
 '608': 35,
 '609': 365,
 '610': 72,
 '612': 303,
 '613': 413,
 '614': 339,
 '615': 110,
 '616': 45,
 '617': 176,
 '621': 92,
 '622': 114,
 '623': 77,
 '624': 54,
 '625': 60,
 '626': 96,
 '627': 286,
 '628': 153,
 '630': 166,
 '631': 258,
 '633': 364,
 '634': 235,
 '635': 106,
 '636': 55,
 '637': 80,
 '638': 146,
 '639': 230,
 '640': 122,
 '641': 320,
 '642': 205,
 '643': 99,
 '646': 258,
 '647': 175,
 '648': 225,
 '650': 571,
 '651': 193,
 '652': 122,
 '654': 122,
 '656': 138,
 '657': 74,
 '658': 160,
 '659': 220,
 '660': 112,
 '661': 192,
 '663': 244}

### 소요시간의 평균 값 구하기

In [4]:
import datetime
import time
import pickle
import numpy as np 
with open('./time/Bike_Stop_Time.pickle', 'rb') as fr:
    Time = pickle.load(fr)

In [8]:
temp2 = {}
for k, l in zip(Bike_Stop_ID, range(51)):
    #temp1[str(k)] = min(list(user_loaded[str(k)].values()).remove(0))
    #print(del list(user_loaded[str(k)].values())[l])
    t = list(Time[str(k)].values())
    del t[l]
    t = np.array(t)
    #print(t)
    #print(min(t))
    temp2[str(k)] = np.mean(t)

In [9]:
temp2

{'600': 779.14,
 '601': 940.26,
 '602': 770.18,
 '604': 609.26,
 '605': 1665.8,
 '606': 742.6,
 '607': 837.46,
 '608': 594.24,
 '609': 1038.1,
 '610': 594.78,
 '612': 1433.02,
 '613': 1630.6,
 '614': 1177.36,
 '615': 903.62,
 '616': 596.34,
 '617': 573.26,
 '621': 615.4,
 '622': 784.38,
 '623': 628.12,
 '624': 645.28,
 '625': 633.3,
 '626': 820.28,
 '627': 797.88,
 '628': 593.4,
 '630': 674.3,
 '631': 815.54,
 '633': 1068.2,
 '634': 797.46,
 '635': 849.28,
 '636': 905.78,
 '637': 963.2,
 '638': 768.68,
 '639': 681.84,
 '640': 735.7,
 '641': 977.76,
 '642': 791.54,
 '643': 844.02,
 '646': 824.56,
 '647': 772.06,
 '648': 727.86,
 '650': 1083.86,
 '651': 928.36,
 '652': 659.1,
 '654': 730.36,
 '656': 831.84,
 '657': 675.26,
 '658': 634.28,
 '659': 1106.14,
 '660': 822.32,
 '661': 857.68,
 '663': 754.36}

In [16]:
temp3 = []
for k in temp2.values():
    temp3 += [k]
    
avg_time = np.mean(np.array(temp3))
avg_time

836.9686274509804

In [15]:
print(datetime.timedelta(0, avg_time))

0:13:56.968627


### 소요시간 정리
- 최댓값
    - 2814 sec
    - 0:46:54
    - 613(신설동역 10번출구 앞) ~ 607(신이문역 2번출구) 대여소
- 최솟값
    - 35 sec
    - 0:00:35
    - 608(전농삼성아파트) ~ 616(시립대 앞) 대여소
- 평균
    - 836.9686274509804 sec
    - 0:13:56.968627
    - 그냥 15분으로 생각하기로 함

In [20]:
for idx in Time['608'].values():
    if idx == 35:
        idx2 = Time['608']
        print(idx2)

{'600': 1201, '601': 576, '602': 685, '604': 426, '605': 706, '606': 859, '607': 1104, '608': 0, '609': 746, '610': 474, '612': 859, '613': 1509, '614': 688, '615': 632, '616': 35, '617': 153, '621': 170, '622': 391, '623': 93, '624': 342, '625': 360, '626': 813, '627': 440, '628': 714, '630': 638, '631': 540, '633': 551, '634': 762, '635': 491, '636': 424, '637': 418, '638': 211, '639': 562, '640': 394, '641': 569, '642': 370, '643': 793, '646': 835, '647': 1395, '648': 698, '650': 941, '651': 282, '652': 426, '654': 227, '656': 327, '657': 222, '658': 663, '659': 658, '660': 720, '661': 421, '663': 1198}
